src/gateio_new_coins_announcements_bot/rotating_proxy.py

```diff
@@ -0,0 +1,108 @@
+import itertools
+import socket
+import struct
+import threading
+import time
+from typing import Callable
+
+import requests
+
+import gateio_new_coins_announcements_bot.globals as globals
+from gateio_new_coins_announcements_bot.logger import logger
+
+
+_proxy_list = {}
+_proxy = None
+_event = threading.Event()
+
+
+def init_proxy():
+    threading.Thread(target=lambda: _every(60 * 10, _fetch_proxies)).start()
+    # Required for populating the proxy list when starting bot
+    _fetch_proxies()
+
+
+def _fetch_proxies():
+    logger.info("Fetching proxies...")
+    _proxy_list
+    global _proxy
+    threads: list[threading.Thread] = []
+    try:
+        proxy_res = requests.get(
+            "https://www.proxyscan.io/api/proxy?last_check=180&limit=20&type=socks5&format=txt&ping=1000"
+        ).text
+    except requests.exceptions.RequestException as e:
+        logger.error(f"Can't fetch proxies. Reason: {e}")
+        return
+
+    # Merging old proxies with new ones
+    _merged_proxies = list(proxy_res[:-1].split("\n") | _proxy_list.keys())
+
+    if len(_merged_proxies) > 0:
+        for p in _merged_proxies:
+            t = threading.Thread(target=checker, args=[p])
+            t.start()
+            threads.append(t)
+
+        for t in threads:
+            t.join()
+
+    logger.info(f"Fetched {len(_proxy_list)} proxies")
+    _proxy = itertools.cycle(_proxy_list.keys())
```

Comment:
- [ ] After some testing i sometimes get an `dictionary changed size during iteration` error (see [here](https://stackoverflow.com/questions/11941817/how-to-avoid-runtimeerror-dictionary-changed-size-during-iteration-error)).

Which is probably because get_proxy() gets called by the main function while you are editing _proxy_list (see `_proxy_list.pop(...)` calls in the `checker`-thread.

We should probably create a copy of the _procy_list and use that to cycle through.
e.g. `_proxy = itertools.cycle(_proxy_list.copy().keys())`
So while we are updating the list, the main thread can still work on the old copy.